### PR TITLE
search-buffer.lisp: tweak search color.

### PR DESCRIPTION
**Pull request by jmercouris**
# Description

This PR changes the search color to be something not confused with a hyperlink. This yellow will need to be added to the theme library as an "attention" color. For now, it is just hacked in there so that you can see what it looks like.

### CODE REVIEW

File: source/mode/search-buffer.lisp [hunk 1]
```diff
@@ -17,19 +17,7 @@
...
-        :animation-iteration-count "infinite"
-        :animation-direction "alternate"
-        :background ,(format nil
-                             "linear-gradient(0deg, ~a 0% 35%, ~a 49% 51%, ~a 65% 100%)"
-                             theme:accent theme:background-alt theme:accent)
-        :background-size 200% 200%)
-      `(:keyframes "mark"
-                   (0% :background-position 0 0%)
-                   (100% :background-position 0 100%)))
+        :background ,(str:concat "#FFEF00" " !important")))
```
Comment (aartaka) [inline source/mode/search-buffer.lisp:20]: You can use the color and `!important` verbatim, LASS allows that:
``` lisp
:background "#FFEF00" !important
```

**Review by aartaka — COMMENTED** [review]
Stylistic remarks aside, I don't agree with this change:
- Yes, links and search matches should be distinguished from each other.
- Color is one of distunguishers, yes.
- But what if links on the page are the same color as search matches, bright yellow?

That's a hypothetical example, but the one that highlights the problem: color alone is not enough to distinguish UI elements (see the discussion in #2983 and #2719). You have to add contrast (contrasting border), texture, or animation (as it used to be before this change). So, whatever the color it is you add, removing the animation is a regression.

----------------------------------------------------------------

I'm not even mentioning the fact that you're adding a color that is beyond our palette and has no solid use-cases for adding it there. We have warning color already, why not use it?—it's bright and attention-catching, after all.

**Comment by aartaka** [comment]
> > ```
> > * But what if links on the page are the same color as search matches, bright yellow?
> > ```
> 
> This is true of any color that we pick.

Why change it to some other arbitrary color then? It's goint to match anyway. Maybe it's not the color that's problematic?

> We are not trying to make the literally most accessible browser in the world. We would completely sacrifice modern aesthetics if we were. Nyxt's mission is: to be a browser for information extraction, for hackers, for professionals.

It's not about accessibility or aesthetics, it's about this PR not having the stated effect.

> > I'm not even mentioning the fact that you're adding a color that is beyond our palette and has no solid use-cases for adding it there. We have warning color already, why not use it?—it's bright and attention-catching, after all.
> 
> This is a solid use case. Highlighting something. Highlighters are typically yellow. We are highlighting matches. Probably we will use highlighting in other places where we wish to highlight text.

We are not trying to make the literally most typical browser in the world. We would completely sacrifice our uniqueness and value proposition if we were. Nyxt's mission is...

See? Rhetorics is cool, but you still haven't rebuked my critique: _any_ color matches at least some of the page elements (due to the Internet using all the possible colors it has as an entity), so repainting search hints into _any_ color does not solve the problem of this color being used by someone else.

**Comment by aartaka** [comment]
> For cases where accessibility is important, we can have a "accessibility" mode. In fact, we can have a "accessible" theme.

I can see where you're coming from, and I know that this PR is likely not _the_ place for this discussion, but: accessibility is important at all times. 
- In ten years, one'll have a weaker eyesight and lower contrast sensitivity, so interfaces with small elements and low-contrast theme will become illegible.
  - In other words, accessibility is important—if not now then a bit later. Get prepared.
- On a sunny day, the screen of a laptop is lit with sunlight. If you have a poor contrast theme, it's all mixed up and illegible.
  - In other words, there's not only a permanent disability, there's a situational one. And there are lots of situations that limit your abilities and perceptions.
- [Up to twenty per cent of males have some form of color blindness](https://www.colour-blindness.com/general/prevalence/),  [Autism Spectrum Disorders cover _at least_ two per cent of population](https://www.cdc.gov/mmwr/volumes/69/ss/ss6904a1.htm), 20/20 vision is a minority. 
  - In other words, accessibility is not about someone "special", it's about a major part of population. And likely, a major part of our users too, because those that are interested in "knowledge extraction software" are quite likely to:
    - Have a weaker eyesight and a certain degree of color blindness due to prolonged screen time.
    - Have an [RSI](https://www.emacswiki.org/emacs/RepeatedStrainInjury) from computer mouse/keyboard interaction.
    - Be an Autism Spectrum Disorder-diagnosed—[programmers are four times more likely to have ASD, for instance](https://developerpitstop.com/autistic-programmers/).
    - etc.

Accessibility is not an afterthought, it's the default. That's why we work on themes with @aadcg. That's why I'm trying to change styles of our UI elements to clealy show their states (focus, click, etc.) That's why we're **not using <div>s** unless necessary.

**Comment by jmercouris** [comment]
> Why change it to some other arbitrary color then? It's goint to match anyway. Maybe it's not the color that's problematic?

Because the color *does* have meaning. I don't agree that the color doesn't have meaning. Yellow is the standard color for highlighting things. Yes, it will collide on a very small amount of pages that have bright yellow backgrounds.

> It's not about accessibility or aesthetics, it's about this PR not having the stated effect.

This is a difference of opinion. I believe that changing the color to yellow will have the effect of focusing the user's eyes.

> We are not trying to make the literally most typical browser in the world. We would completely sacrifice our uniqueness and value proposition if we were. Nyxt's mission is...
> 
> See? Rhetorics is cool, but you still haven't rebuked my critique: _any_ color matches at least some of the page elements (due to the Internet using all the possible colors it has as an entity), so repainting search hints into _any_ color does not solve the problem of this color being used by someone else.

I'm not sure what you are trying to say here. I am just stating that color accessibility is not within Nyxt's core mission. We won't eschew it, but we won't bend over backwards to make it happen when it takes away from the other goals of the project.

**Comment by jmercouris** [comment]
Finally, that color yellow with the color black *does* meet the minimum contrast standards.

**Comment by Ambrevar** [comment]
I believe this color change was specified by @lansingthomas, right?

In any case, @aartaka let's not waste too much sweat on this, it's relatively minor (and also fully customizable by the user).  Like John said, we can totally have another theme for improved accessibility.

**Comment by aartaka** [comment]
> I believe this color change was specified by @lansingthomas, right?

I have (yet anther?) radical proposition for theming and colors: let's scrape the theming from Bootstrap _and freeze the theming API_. Not forever, but for a long enough time—say a year. Because trying to solve all the problems by adding Yet Another Color™️ is a pure escapism from actually designing interfaces that are flexible to the context they're used in.

> In any case, @aartaka let's not waste too much sweat on this, it's relatively minor (and also fully customizable by the user). 

By "minor" you mean minor version bump for slight API breakage? :P

> Like John said, we can totally have another theme for improved accessibility.

Uh oh, I guess my comment about accessibility didn't get posted (GitHub CRDTs unreliability?) :P

Jokes aside, my complains are not about accessibility, they are about Yet Another Color™️ gold rush and it's naïvety.

**Review by aadcg — CHANGES_REQUESTED** [review]
For background, the previous style for search matches was designed by me and @jmercouris. It happened like that because @jmercouris  was the only active reviewer of #2908. I then asked for @lansingthomas to review the UI side of it (the styling), since I had doubts about the animation strategy. Plus, @lansingthomas is the UI gatekeeper.

--------

The topic of accessibility has been brought up, and I think it is mildly important indeed. But everyone failed to make one really simple and obvious observation. Before #2908 landed, here's a snippet of the styling:

```lisp 
(".nyxt-search-node > .nyxt-hint.nyxt-select-hint"
        :background-color ,theme:accent
        :color ,theme:on-accent)
```

Can you see the accessibility argument? If you set a background color, you need to set the (foreground) color as well. This PR missed this detail, and that's actually the only (minor) issue I have with it. It should hardcore black as the (foreground) color as to ensure that the yellow marker is used over black text. 

-----

I'd like to second what @Ambrevar has said. 

> Let's not waste too much sweat on this, it's relatively minor (and also fully customizable by the user).

Also, I think it's important to stay on topic. Some decisions just come down to preference. And, in these cases, voting is more efficient. I'm actually OK with the old style and the new one (granted that my observation above is met).  

But my vote goes to going forward with PR since I like when people hear their hats (and "suffer" the consequences of responsibility). If a user tomorrow constructively arguments that the search mark styling sucks, then I know who to "blame". Right @lansingthomas? :D

Please cast your votes. Thanks!

**Comment by lansingthomas** [comment]
>I believe this color change was specified by @lansingthomas, right?

## how we got stuck

The idea and implementation were a group thing, I should have opened this as an #Issue after prototyping with john. MY BAD LADS.

John and Andre were going back and forth on the highlighting animation, they asked me to weigh in. I do not like the gradient animation at all for this so wanted to provide a better solution. Andre and I looked at a number of other apps and how they handle this, then elaborated some user expectations. I then asked John to prototype it (easier than mockups in this case). 

The specific yellow came from our  (@aartaka and me) discussion about warning colors. We chose this and another yellow as a draft for the _warning_, and _warning-alt_ colors in our palette before settling on red (more recognizable as a warning than yellow).

Screenshot 2023-05-31 at 3 43 15 PM

I assumed (like a donkey :P ) the yellow would be fine because it:
[1] meets our agreed on contrast standards, [2] fits the palette Artyom and I laid out, [3] connects with the physical yellow of a highlighter marker, [4] mirrors Chrome-browser's yellow, [5] avoids negative interactions with hyperlink blue, and [6] avoids potential future negative interactions with blue text in our own documentation (not yet announced or agreed on but a possibility nonetheless).

## Understanding that;
@artyom is right about this:
>color alone is not enough to distinguish UI elements...

we must add other things to be good builders for all.
and
the animation was fun, but not perfect.

SO

## I propose this course of action to get un-stuck

I propose that we do BOTH (color change) and (form factor change of some kind)

for the color change: accept the PR and go for yellow
for the form factor change: I open another issue and we can discuss some other form factors. (including revisiting color if it seems to make sense)


Possibilities

**Comment by lansingthomas** [comment]
haha we are both writing essays to each other here andre :P

**Comment by Ambrevar** [comment]
> I have (yet anther?) radical proposition for theming and colors: let's scrape the theming from Bootstrap _and freeze the theming API_. Not forever, but for a long enough time—say a year. Because trying to solve all the problems by adding Yet Another Colortm is a pure escapism from actually designing interfaces that are flexible to the context they're used in.

Sounds fair!



> > In any case, @aartaka let's not waste too much sweat on this, it's relatively minor (and also fully customizable by the user).
> 
> By "minor" you mean minor version bump for slight API breakage? :P
> 
> > Like John said, we can totally have another theme for improved accessibility.
> 
> Uh oh, I guess my comment about accessibility didn't get posted (GitHub CRDTs unreliability?) :P

Please, do not escalate @aartaka.

I didn't ignore your accessibility comment, the opposite in fact: I meant to underline that we should take it into account.  But that can be done in another iteration.

Theming a browser in Common Lisp is evidently hard, and we're currently struggling not only because it's hard but also because the discussions around it are not very smooth to say the least.  So let's try to be a bit more lenient about it, lest we walk in circles and get nowhere.

@aartaka @lansingthomas Do you want to start a new discussion about a relatively-definitive theming API based on Bootstrap?

**Comment by aadcg** [comment]
> @aartaka @lansingthomas Do you want to start a new discussion about a relatively-definitive theming API based on Bootstrap?

@Ambrevar I know you're proposing it with the best of intentions, but I doubt @lansingthomas is interested in engaging in such endeavors. Recently, there has been a thoughtful and long re-iteration of the themes (see #2893). From @lansingthomas's perspective the conclusion is "I had to sweat blood to get a new color in... :open_mouth:". It's quite frustrating for a person whose responsibility is to improve the UI. Experimentation is key for improvement.

Let me stress the fact that it's an unfair "fight" when @lansingthomas tries to win his space among 4 engineers + being the youngest team member. I'm playing to be his PR assistant since I don't think we try to put ourselves in his shoes enough, and he will always be too politically correct to make us realize.


-----

I think we're raising the question: is it OK to use a color (in this case, yellow) that isn't part of the Nyxt theme palette? 

For me the answer is yes. Recall that we had the following style for `diff-mode`:

```lisp
(theme:themed-css (theme *browser*)
            `(".nyxt-diff-insert"
              :background-color "#bbeabb"
              :text-decoration "none")
            `("ins.nyxt-diff-replace"
              :background-color "#bbeabb"
              :text-decoration "none")
            `(".nyxt-diff-delete"
              :background-color "#efcbcf"
              :text-decoration "none")
            `("del.nyxt-diff-replace"
              :background-color "#efcbcf"
              :text-decoration "none"))
```

**Comment by aartaka** [comment]
> Let me stress the fact that it's an unfair "fight" when @lansingthomas tries to win his space among 4 engineers + being the youngest team member.

Wait, really? There's someone on the team that's younger than myself? :D

**Comment by aartaka** [comment]
All in all:
- I disagree with adding yellow, because all the agruments in favor of it ("typical color", "was in one of palette iterations", "avoiding clashes") are either not value-backed or are addressable with other better solutions.
- The old animation I don't like, the new one I haven't seen yet. But here I'm for any animation, because it's better than none.
- I disagree with addressing the accessibility/texture/non-color affordance later, because postponing it beyond _this PR_ ™️ is how we forget about things.

So what I'm suggesting here is:
- Retaining the accent color.
- Experimenting with texture/border/animation etc.
- And adding a new color _only_ if these are severely insufficient to:
  - Discern links and search marks.
  - Draw one's attention.
  - ~Have a Google-like typical browser corporate design look :P~

**Comment by lansingthomas** [comment]
>Wait, really? There's someone on the team that's younger than myself? :D

I may have a young spirit, though I certainly have some more years :)
I think Andre means that I have less time in Nyxt.

#### solution space
I notice that we agree that the form factor can be improved. Let's use that to move this forward.
- I will open an #Issue immediately. See you there :D
- The color is out of theme, so fine. **I accept the blue.**

@aartaka let's have a conversation about Bootstrap based relatively-definitive theming on Monday.
- I propose Monday in the new Jitsi at our typical time
- anyone else feel free to join
- 17:00 Monday UTC

- [ ] cool with you?

#### some happy pushback
>agruments in favor of it ("typical color", "was in one of palette iterations", "avoiding clashes") are either not value-backed... 

Alright... You are right to stand up against colors outside of our official theme, though my arguments are certainly value backed. Give me some credit here man. I don't want to fight with you about this (or anything <3).

Argument | **Value**
[1] meets our agreed on contrast standards, | **VALUE**: accessibility, error prevention, constraints
[2] fits the palette Artyom and I laid out, | **VALUE**: consistency, sense of place in system
[3] connects with the physical yellow of a highlighter marker, **VALUE**: consistency across apps and physical space, matching mental models
[4] mirrors Chrome-browser's yellow, **VALUE**: Steal Like An Artist, consistency, matching
[5] avoids negative interactions with hyperlink blue, **VALUE**: progressive disclosure, error prevention, anticipation of user needs, appropriately minimal aesthetics
[6] avoids potential future negative interactions with blue text in our own documentation (not yet announced or agreed on but a possibility nonetheless). **VALUE**: error prevention

Yay building nice things!

**Comment by lansingthomas** [comment]
>Experimenting with texture/border/animation etc.

Some ideas for your experimentation to replace the gradient animation; (no need to respond to all of them -- just ideas)
- thin border around the selected hint highlights (_accent-alt_ perhaps)
- see how things look with and without the corner rounding
- add two new colors (don't kill me): I notice that `accent blue`, `hypothetical-warning-yellow`, and `super-hypothetical-red/violet` create a nice triadic color which could look nice. (accent for the Suggestion in the prompt buffer, and the other two hypothetical colors for the hints {selected, unselected})
- float a little asterisk or triangle in the corner of the selected hint highlight.
- create a fat underline in the highlight layer by changing the hint highlight color between baseline and descender height to _accent-alt_
- change the unselected hint highlight color to a lighter grey from our palette to emphasize the selected hint.

**Comment by lansingthomas** [comment]
Update: Artyom and I had a productive talk yesterday that should solve this problem (and many others!) by creating a more comprehensive theme / palette. It is WIP and we are going to get some more official colors. *commence international celebration* : )

**Comment by jmercouris** [comment]
I look forward to the official palette!

**Comment by lansingthomas** [comment]
sorry I am dragging on the palette that is blocking this one -- still WIP. 

patience is appreciated.

**Comment by jmercouris** [comment]
Patience is had ;-)

**Comment by lansingthomas** [comment]
Today is the day!
for highlighter colors.

# light theme
highlight colors light theme (2)


# dark theme
highlight colors dark theme (1)

Good enough?
YES
- [ ] @aadcg 
- [ ] @jmercouris 
- [ ] @aartaka 
- [ ] @Ambrevar 

NO
- [ ] @aadcg 
- [ ] @jmercouris 
- [ ] @aartaka 
- [ ] @Ambrevar 


### a special note for the hero who takes on the implementation 
- a nice structure for color variables can be found in #3005 
- please let me know if unforeseen interaction states occur (_maybe with links or something_). I trust you to make good choices but I am tuning our interaction guidelines so information is helpful at this stage.

### Observations for posterity
- having tried a number of color combinations, and understanding there are always tradeoffs... I believe these are the best we have.
- Yes, I see that the highlighted link text has poor contrast. There are a few fixes rattling around in my brain, none of them involve changing the highlighted text color.

**Pull request closed without merging**